src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs

```diff
@@ -22,24 +22,39 @@ public ForLoopBinder(Binder enclosing, ForStatementSyntax syntax)
 
         override protected ImmutableArray<LocalSymbol> BuildLocals()
         {
+            var locals = ArrayBuilder<LocalSymbol>.GetInstance();
+
             var declaration = _syntax.Declaration;
-            if (declaration == null)
+            if (declaration != null)
             {
-                return ImmutableArray<LocalSymbol>.Empty;
-            }
+                var refKind = _syntax.RefKeyword.Kind().GetRefKind();
 
-            var refKind = _syntax.RefKeyword.Kind().GetRefKind();
+                foreach (var variable in declaration.Variables)
+                {
+                    var localSymbol = MakeLocal(refKind,
+                                                declaration,
+                                                variable,
+                                                LocalDeclarationKind.ForInitializerVariable);
+                    locals.Add(localSymbol);
 
-            var locals = ArrayBuilder<LocalSymbol>.GetInstance();
-            foreach (var variable in declaration.Variables)
+                    if (variable.Initializer != null)
+                    {
+                        BuildAndAddPatternVariables(locals, variable.Initializer.Value);
+                    }
+                }
+            }
+            else
             {
-                var localSymbol = MakeLocal(refKind,
-                                            declaration,
-                                            variable,
-                                            LocalDeclarationKind.ForInitializerVariable);
-                locals.Add(localSymbol);
+                BuildAndAddPatternVariables(locals, nodes:_syntax.Initializers.ToImmutableArray<CSharpSyntaxNode>());
```

Comment:
This uses the `IEnumerable` `ToImmutableArray` extension method right? Does that use a pooled builder? Should we consider implementing an extension method specifically for `SeparatedSyntaxList`?